File: lab4/samples/memory_profiler.py
#!/usr/bin/env python3
"""
memory_profiler.py - Профилировщик памяти процессов (Python версия)

Использование:
    python3 memory_profiler.py <PID> [--watch] [--interval SECONDS]

Альтернативная реализация на Python для тех, кто предпочитает Python вместо C.
Использует библиотеку psutil для более простого доступа к метрикам.

Требования:
    pip3 install psutil
"""

import sys
import time
import argparse
from pathlib import Path

try:
    import psutil
except ImportError:
    print("Error: psutil library not found.")
    print("Install it with: pip3 install psutil")
    sys.exit(1)


class MemoryProfiler:
    """Профилировщик памяти для процесса"""

    def __init__(self, pid):
        self.pid = pid
        try:
            self.process = psutil.Process(pid)
        except psutil.NoSuchProcess:
            print(f"Error: Process {pid} does not exist")
            sys.exit(1)
        except psutil.AccessDenied:
            print(f"Error: Access denied to process {pid}")
            sys.exit(1)

    def get_memory_info(self):
        """Получить базовую информацию о памяти"""
        # TODO: Реализовать получение метрик через psutil
        # mem_info = self.process.memory_info()
        # mem_full = self.process.memory_full_info()
        #
        # return {
        #     'vms': mem_info.vms,      # Virtual Memory Size
        #     'rss': mem_info.rss,      # Resident Set Size
        #     'shared': mem_full.shared,
        #     'text': mem_full.text,
        #     'data': mem_full.data,
        #     'lib': mem_full.lib,
        #     'dirty': mem_full.dirty,
        #     'uss': mem_full.uss,      # Unique Set Size
        #     'pss': mem_full.pss       # Proportional Set Size
        # }

        return {}

    def get_page_faults(self):
        """Получить статистику page faults"""
        # TODO: Использовать psutil для получения page faults
        # Подсказка: process.num_page_faults()
        return {'minor': 0, 'major': 0}

    def get_memory_maps(self):
        """Получить карту памяти процесса"""
        # TODO: Использовать process.memory_maps()
        # Это возвращает список объектов с полями:
        # - path: путь к файлу или [heap], [stack]
        # - rss: размер в памяти
        # - size: виртуальный размер
        # - pss: proportional set size
        # - shared_clean, shared_dirty, private_clean, private_dirty

        return []

    def print_basic_info(self):
        """Вывести базовую информацию о процессе"""
        print(f"Process: {self.process.name()} (PID {self.pid})")
        print("=" * 60)
        print()

        # TODO: Получить и вывести информацию о памяти
        # mem = self.get_memory_info()

        # print("Memory Metrics:")
        # print(f"  VSZ (Virtual):     {self._format_size(mem['vms'])}")
        # print(f"  RSS (Resident):    {self._format_size(mem['rss'])}")
        # print(f"  PSS (Proportional):{self._format_size(mem['pss'])} (more accurate)")
        # print(f"  USS (Unique):      {self._format_size(mem['uss'])}")
        # print()
        # print(f"  Shared memory:     {self._format_size(mem['shared'])}")
        # print(f"  Text (code):       {self._format_size(mem['text'])}")
        # print(f"  Data + Heap:       {self._format_size(mem['data'])}")
        # print(f"  Libraries:         {self._format_size(mem['lib'])}")
        # print()

        # TODO: Вывести page faults
        # faults = self.get_page_faults()
        # print("Page Faults:")
        # print(f"  Minor: {faults['minor']}")
        # print(f"  Major: {faults['major']}")

    def print_memory_map(self, limit=20):
        """Вывести карту памяти"""
        print()
        print("Memory Map:")
        print(f"{'Path':<40} {'RSS':>12} {'PSS':>12} {'Size':>12}")
        print("-" * 80)

        # TODO: Получить и вывести карту памяти
        # maps = self.get_memory_maps()
        #
        # # Сгруппировать по типу
        # for i, m in enumerate(maps[:limit]):
        #     print(f"{m.path[:40]:<40} "
        #           f"{self._format_size(m.rss):>12} "
        #           f"{self._format_size(m.pss):>12} "
        #           f"{self._format_size(m.size):>12}")
        #
        # if len(maps) > limit:
        #     print(f"... ({len(maps) - limit} more entries)")

    def watch(self, interval=1):
        """Мониторинг процесса в реальном времени"""
        print(f"Monitoring PID {self.pid} (update every {interval}s, Ctrl+C to stop)")
        print()

        prev_mem = None
        prev_faults = None

        try:
            while True:
                print("\n" + "=" * 60)
                print(f"Time: {time.strftime('%Y-%m-%d %H:%M:%S')}")
                print(f"Process: {self.process.name()} (PID {self.pid})")
                print()

                # TODO: Получить текущие метрики
                # mem = self.get_memory_info()
                # faults = self.get_page_faults()

                # TODO: Вывести с дельтами
                # print(f"VSZ: {self._format_size(mem['vms'])}", end="")
                # if prev_mem:
                #     delta = mem['vms'] - prev_mem['vms']
                #     if delta != 0:
                #         print(f"  ({self._format_delta(delta)})", end="")
                # print()
                #
                # ... аналогично для других метрик
                #
                # prev_mem = mem
                # prev_faults = faults

                time.sleep(interval)

        except KeyboardInterrupt:
            print("\n\nMonitoring stopped.")

    def _format_size(self, bytes_val):
        """Форматировать размер в человекочитаемый вид"""
        # TODO: Реализовать форматирование
        # KB = 1024
        # MB = KB * 1024
        # GB = MB * 1024
        #
        # if bytes_val < KB:
        #     return f"{bytes_val} B"
        # elif bytes_val < MB:
        #     return f"{bytes_val/KB:.1f} KB"
        # elif bytes_val < GB:
        #     return f"{bytes_val/MB:.1f} MB"
        # else:
        #     return f"{bytes_val/GB:.2f} GB"

        return f"{bytes_val} B"

    def _format_delta(self, delta):
        """Форматировать изменение"""
        sign = "+" if delta > 0 else ""
        return f"{sign}{self._format_size(abs(delta))}"


def compare_processes(pid1, pid2):
    """Сравнить два процесса"""
    print(f"Comparing processes: {pid1} vs {pid2}")
    print("=" * 60)
    print()

    # TODO: Создать два профайлера и сравнить метрики
    # profiler1 = MemoryProfiler(pid1)
    # profiler2 = MemoryProfiler(pid2)
    #
    # mem1 = profiler1.get_memory_info()
    # mem2 = profiler2.get_memory_info()
    #
    # print(f"{'Metric':<20} {'PID ' + str(pid1):>15} {'PID ' + str(pid2):>15} {'Difference':>15}")
    # print("-" * 70)
    #
    # for key in ['vms', 'rss', 'pss', 'uss']:
    #     diff = mem2[key] - mem1[key]
    #     print(f"{key.upper():<20} "
    #           f"{profiler1._format_size(mem1[key]):>15} "
    #           f"{profiler2._format_size(mem2[key]):>15} "
    #           f"{profiler1._format_delta(diff):>15}")


def main():
    parser = argparse.ArgumentParser(
        description="Memory profiler for Linux processes (Python version)",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s 1234                    Show info for PID 1234
  %(prog)s 1234 --watch            Monitor PID 1234
  %(prog)s 1234 --watch -i 5       Monitor with 5 sec interval
  %(prog)s 1234 --compare 5678     Compare two processes
  %(prog)s 1234 --map              Show detailed memory map
        """
    )

    parser.add_argument('pid', type=int, help='Process ID to analyze')
    parser.add_argument('--watch', action='store_true',
                        help='Monitor process continuously')
    parser.add_argument('-i', '--interval', type=int, default=1,
                        help='Update interval in seconds (default: 1)')
    parser.add_argument('--compare', type=int, metavar='PID2',
                        help='Compare with another process')
    parser.add_argument('--map', action='store_true',
                        help='Show detailed memory map')

    args = parser.parse_args()

    # Режим сравнения
    if args.compare:
        compare_processes(args.pid, args.compare)
        return

    # Создать профайлер
    profiler = MemoryProfiler(args.pid)

    # Режим мониторинга
    if args.watch:
        profiler.watch(args.interval)
    else:
        # Одиночный вывод
        profiler.print_basic_info()

        if args.map:
            profiler.print_memory_map()


if __name__ == '__main__':
    main()


"""
ЗАДАНИЯ для студента:

1. Установите psutil:
   $ pip3 install psutil

2. Реализуйте все TODO функции:
   - get_memory_info()
   - get_page_faults()
   - get_memory_maps()
   - print_basic_info()
   - print_memory_map()
   - watch()
   - _format_size()

3. Протестируйте на разных процессах:
   $ python3 memory_profiler.py $$
   $ python3 memory_profiler.py 1
   $ firefox &
   $ python3 memory_profiler.py $(pgrep firefox) --watch

4. Сравните результаты с C-версией и системными утилитами:
   $ ./memory_profiler 1234
   $ python3 memory_profiler.py 1234
   $ ps -o pid,vsz,rss -p 1234

5. Реализуйте дополнительные возможности:
   - График изменения памяти (используя matplotlib)
   - Сохранение данных в CSV
   - Поиск процессов по имени
   - Интерактивный режим (выбор из списка процессов)

6. Создайте визуализацию:
   $ python3 memory_profiler.py 1234 --watch --graph
   (показывать ASCII-график изменения RSS в реальном времени)

ПРЕИМУЩЕСТВА Python версии:
- Проще в реализации (psutil делает всю работу)
- Кросс-платформенность (работает на Windows, macOS, Linux)
- Легко добавить визуализацию (matplotlib, plotly)
- Интерактивность (rich library для красивого вывода)

НЕДОСТАТКИ:
- Зависимость от внешних библиотек
- Может быть медленнее для большого количества процессов
- Требует интерпретатор Python

Выберите C или Python версию в зависимости от ваших предпочтений!
"""

File: lab4/gr9-sub1_lab4_RozhdaevEV/src/memory_profiler.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <dirent.h>
#include <time.h>
#include <errno.h>
#include <signal.h>

#define MAX_LINE_LENGTH 1024
#define MAX_MAPS_ENTRIES 1000
#define WATCH_INTERVAL_DEFAULT 2

typedef struct {
    unsigned long vms;    
    unsigned long rss;  
    unsigned long pss;    
    unsigned long uss;  
    unsigned long shared; 
    unsigned long text;   
    unsigned long data;  
    unsigned long lib;    
    long minor_faults;    
    long major_faults;     
} MemoryMetrics;

typedef struct {
    unsigned long start;
    unsigned long end;
    char perms[8];
    unsigned long offset;
    char dev[16];
    unsigned long inode;
    char pathname[256];
    unsigned long size;
    unsigned long rss;
    unsigned long pss;
} MemoryMapEntry;

typedef struct {
    pid_t pid;
    char name[256];
    MemoryMetrics metrics;
    MemoryMapEntry* maps;
    int maps_count;
} ProcessInfo;

int read_memory_metrics(pid_t pid, MemoryMetrics* metrics);
int read_memory_maps(pid_t pid, MemoryMapEntry** maps, int* count);
void free_memory_maps(MemoryMapEntry* maps, int count);
int read_process_name(pid_t pid, char* name, size_t name_len);

void print_basic_info(const ProcessInfo* proc);
void print_memory_map(const ProcessInfo* proc, int limit);
void print_comparison(const ProcessInfo* proc1, const ProcessInfo* proc2);
void print_watch_header(void);

const char* format_size(unsigned long bytes);
const char* format_delta(long delta, char* buffer, size_t buffer_size);
double get_time(void);
void clear_screen(void);
int parse_arguments(int argc, char* argv[], pid_t* pid1, pid_t* pid2,
    int* watch_mode, int* interval, int* show_map);

static volatile int keep_running = 1;

void signal_handler(int sig) {
    keep_running = 0;
    printf("\nStopping monitoring...\n");
}

int read_memory_metrics(pid_t pid, MemoryMetrics* metrics) {
    char path[256];
    FILE* f;
    char line[MAX_LINE_LENGTH];

    memset(metrics, 0, sizeof(MemoryMetrics));

    snprintf(path, sizeof(path), "/proc/%d/status", pid);
    f = fopen(path, "r");
    if (!f) {
        return -1;
    }

    while (fgets(line, sizeof(line), f)) {
        if (strncmp(line, "VmSize:", 7) == 0) {
            sscanf(line, "VmSize: %lu kB", &metrics->vms);
            metrics->vms *= 1024;  
        }
        else if (strncmp(line, "VmRSS:", 6) == 0) {
            sscanf(line, "VmRSS: %lu kB", &metrics->rss);
            metrics->rss *= 1024;
        }
        else if (strncmp(line, "VmData:", 7) == 0) {
            sscanf(line, "VmData: %lu kB", &metrics->data);
            metrics->data *= 1024;
        }
        else if (strncmp(line, "VmStk:", 6) == 0) {
        }
        else if (strncmp(line, "VmExe:", 6) == 0) {
            sscanf(line, "VmExe: %lu kB", &metrics->text);
            metrics->text *= 1024;
        }
        else if (strncmp(line, "VmLib:", 6) == 0) {
            sscanf(line, "VmLib: %lu kB", &metrics->lib);
            metrics->lib *= 1024;
        }
        else if (strncmp(line, "Name:", 5) == 0) {
        }
    }
    fclose(f);

    snprintf(path, sizeof(path), "/proc/%d/stat", pid);
    f = fopen(path, "r");
    if (f) {
        for (int i = 0; i < 9; i++) {
            if (fscanf(f, "%*s") == EOF) break;
        }
        fscanf(f, "%ld %*s %ld", &metrics->minor_faults, &metrics->major_faults);
        fclose(f);
    }

    snprintf(path, sizeof(path), "/proc/%d/smaps_rollup", pid);
    f = fopen(path, "r");
    if (f) {
        while (fgets(line, sizeof(line), f)) {
            if (strncmp(line, "Pss:", 4) == 0) {
                unsigned long pss_kb;
                sscanf(line, "Pss: %lu kB", &pss_kb);
                metrics->pss = pss_kb * 1024;
            }
            else if (strncmp(line, "Private_Clean:", 14) == 0 ||
                strncmp(line, "Private_Dirty:", 14) == 0) {
                unsigned long private_kb;
                sscanf(line, "%*s %lu kB", &private_kb);
                metrics->uss += private_kb * 1024;
            }
            else if (strncmp(line, "Shared_Clean:", 13) == 0 ||
                strncmp(line, "Shared_Dirty:", 13) == 0) {
                unsigned long shared_kb;
                sscanf(line, "%*s %lu kB", &shared_kb);
                metrics->shared += shared_kb * 1024;
            }
        }
        fclose(f);
    }
    else {
        metrics->pss = metrics->rss;  
        metrics->uss = metrics->rss / 2;  
    }

    return 0;
}

int read_memory_maps(pid_t pid, MemoryMapEntry** maps, int* count) {
    char path[256];
    FILE* f;
    char line[MAX_LINE_LENGTH];
    MemoryMapEntry* entries = NULL;
    int entry_count = 0;
    int capacity = 0;

    snprintf(path, sizeof(path), "/proc/%d/maps", pid);
    f = fopen(path, "r");
    if (!f) {
        return -1;
    }

    while (fgets(line, sizeof(line), f)) {
        if (entry_count >= capacity) {
            capacity = capacity == 0 ? 50 : capacity * 2;
            MemoryMapEntry* new_entries = realloc(entries, capacity * sizeof(MemoryMapEntry));
            if (!new_entries) {
                free(entries);
                fclose(f);
                return -1;
            }
            entries = new_entries;
        }

        MemoryMapEntry* entry = &entries[entry_count];
        memset(entry, 0, sizeof(MemoryMapEntry));

        if (sscanf(line, "%lx-%lx %7s %lx %15s %lu %255[^\n]",
            &entry->start, &entry->end, entry->perms, &entry->offset,
            entry->dev, &entry->inode, entry->pathname) >= 6) {

            entry->size = entry->end - entry->start;
            entry_count++;
        }
    }

    fclose(f);

    *maps = entries;
    *count = entry_count;
    return 0;
}

void free_memory_maps(MemoryMapEntry* maps, int count) {
    free(maps);
}

int read_process_name(pid_t pid, char* name, size_t name_len) {
    char path[256];
    FILE* f;

    snprintf(path, sizeof(path), "/proc/%d/comm", pid);
    f = fopen(path, "r");
    if (!f) {
        return -1;
    }

    if (fgets(name, name_len, f)) {
        size_t len = strlen(name);
        if (len > 0 && name[len - 1] == '\n') {
            name[len - 1] = '\0';
        }
    }
    else {
        fclose(f);
        return -1;
    }

    fclose(f);
    return 0;
}

const char* format_size(unsigned long bytes) {
    static char buffer[32];
    const char* units[] = { "B", "KB", "MB", "GB" };
    int unit_index = 0;
    double size = bytes;

    while (size >= 1024.0 && unit_index < 3) {
        size /= 1024.0;
        unit_index++;
    }

    if (unit_index == 0) {
        snprintf(buffer, sizeof(buffer), "%lu %s", bytes, units[unit_index]);
    }
    else {
        snprintf(buffer, sizeof(buffer), "%.1f %s", size, units[unit_index]);
    }

    return buffer;
}

const char* format_delta(long delta, char* buffer, size_t buffer_size) {
    if (delta == 0) {
        snprintf(buffer, buffer_size, "0");
    }
    else if (delta > 0) {
        snprintf(buffer, buffer_size, "+%s", format_size(delta));
    }
    else {
        snprintf(buffer, buffer_size, "-%s", format_size(-delta));
    }
    return buffer;
}

double get_time(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

void clear_screen(void) {
    printf("\033[2J\033[H");
}

void print_basic_info(const ProcessInfo* proc) {
    printf("Process: %s (PID %d)\n", proc->name, proc->pid);
    printf("=");
    for (int i = 0; i < 60; i++) printf("=");
    printf("\n\n");

    printf("Memory Metrics:\n");
    printf("  VSZ (Virtual):     %s\n", format_size(proc->metrics.vms));
    printf("  RSS (Resident):    %s\n", format_size(proc->metrics.rss));

    if (proc->metrics.pss > 0) {
        printf("  PSS (Proportional): %s (more accurate)\n", format_size(proc->metrics.pss));
    }
    if (proc->metrics.uss > 0) {
        printf("  USS (Unique):      %s\n", format_size(proc->metrics.uss));
    }

    printf("\n");
    printf("  Shared memory:     %s\n", format_size(proc->metrics.shared));
    printf("  Text (code):       %s\n", format_size(proc->metrics.text));
    printf("  Data + Heap:       %s\n", format_size(proc->metrics.data));
    printf("  Libraries:         %s\n", format_size(proc->metrics.lib));
    printf("\n");

    printf("Page Faults:\n");
    printf("  Minor: %ld\n", proc->metrics.minor_faults);
    printf("  Major: %ld\n", proc->metrics.major_faults);
}

void print_memory_map(const ProcessInfo* proc, int limit) {
    printf("\nMemory Map:");
    if (limit > 0 && proc->maps_count > limit) {
        printf(" (showing first %d of %d entries)\n", limit, proc->maps_count);
    }
    else {
        printf(" (%d entries)\n", proc->maps_count);
    }

    printf("%-18s %-6s %-8s %s\n", "Address Range", "Perms", "Size", "Path");
    printf("-");
    for (int i = 0; i < 78; i++) printf("-");
    printf("\n");

    int show_count = (limit > 0 && proc->maps_count > limit) ? limit : proc->maps_count;

    for (int i = 0; i < show_count; i++) {
        const MemoryMapEntry* entry = &proc->maps[i];
        printf("%08lx-%08lx %-6s %-8s %s\n",
            entry->start, entry->end, entry->perms,
            format_size(entry->size), entry->pathname);
    }

    if (proc->maps_count > show_count) {
        printf("... (%d more entries)\n", proc->maps_count - show_count);
    }
}

void print_comparison(const ProcessInfo* proc1, const ProcessInfo* proc2) {
    printf("Comparing processes: %s (PID %d) vs %s (PID %d)\n",
        proc1->name, proc1->pid, proc2->name, proc2->pid);
    printf("=");
    for (int i = 0; i < 70; i++) printf("=");
    printf("\n\n");

    printf("%-20s %15s %15s %15s\n",
        "Metric",
        proc1->name,
        proc2->name,
        "Difference");
    printf("-");
    for (int i = 0; i < 70; i++) printf("-");
    printf("\n");

    char delta_buffer[32];

    long vsz_delta = proc2->metrics.vms - proc1->metrics.vms;
    printf("%-20s %15s %15s %15s\n",
        "VSZ",
        format_size(proc1->metrics.vms),
        format_size(proc2->metrics.vms),
        format_delta(vsz_delta, delta_buffer, sizeof(delta_buffer)));

    long rss_delta = proc2->metrics.rss - proc1->metrics.rss;
    printf("%-20s %15s %15s %15s\n",
        "RSS",
        format_size(proc1->metrics.rss),
        format_size(proc2->metrics.rss),
        format_delta(rss_delta, delta_buffer, sizeof(delta_buffer)));

    if (proc1->metrics.pss > 0 && proc2->metrics.pss > 0) {
        long pss_delta = proc2->metrics.pss - proc1->metrics.pss;
        printf("%-20s %15s %15s %15s\n",
            "PSS",
            format_size(proc1->metrics.pss),
            format_size(proc2->metrics.pss),
            format_delta(pss_delta, delta_buffer, sizeof(delta_buffer)));
    }

    long minor_delta = proc2->metrics.minor_faults - proc1->metrics.minor_faults;
    printf("%-20s %15ld %15ld %15+ld\n",
        "Minor Faults",
        proc1->metrics.minor_faults,
        proc2->metrics.minor_faults,
        minor_delta);

    long major_delta = proc2->metrics.major_faults - proc1->metrics.major_faults;
    printf("%-20s %15ld %15ld %15+ld\n",
        "Major Faults",
        proc1->metrics.major_faults,
        proc2->metrics.major_faults,
        major_delta);
}

void print_watch_header(void) {
    printf("%-12s %-12s %-12s %-12s %-12s %-12s\n",
        "Time", "VSZ", "RSS", "PSS", "MinorF", "MajorF");
    printf("-");
    for (int i = 0; i < 72; i++) printf("-");
    printf("\n");
}

int parse_arguments(int argc, char* argv[], pid_t* pid1, pid_t* pid2,
    int* watch_mode, int* interval, int* show_map) {
    *pid1 = 0;
    *pid2 = 0;
    *watch_mode = 0;
    *interval = WATCH_INTERVAL_DEFAULT;
    *show_map = 0;

    if (argc < 2) {
        return -1;
    }

    *pid1 = atoi(argv[1]);
    if (*pid1 <= 0) {
        fprintf(stderr, "Error: Invalid PID '%s'\n", argv[1]);
        return -1;
    }

    for (int i = 2; i < argc; i++) {
        if (strcmp(argv[i], "--watch") == 0) {
            *watch_mode = 1;
            if (i + 1 < argc && argv[i + 1][0] != '-') {
                *interval = atoi(argv[i + 1]);
                if (*interval <= 0) *interval = WATCH_INTERVAL_DEFAULT;
                i++;
            }
        }
        else if (strcmp(argv[i], "--compare") == 0 && i + 1 < argc) {
            *pid2 = atoi(argv[i + 1]);
            if (*pid2 <= 0) {
                fprintf(stderr, "Error: Invalid second PID '%s'\n", argv[i + 1]);
                return -1;
            }
            i++;
        }
        else if (strcmp(argv[i], "--map") == 0) {
            *show_map = 1;
        }
        else if (strcmp(argv[i], "--help") == 0) {
            printf("Usage: %s <PID> [OPTIONS]\n\n", argv[0]);
            printf("Options:\n");
            printf("  --watch [INTERVAL]    Monitor process continuously (default: %d sec)\n", WATCH_INTERVAL_DEFAULT);
            printf("  --compare PID2        Compare with another process\n");
            printf("  --map                 Show detailed memory map\n");
            printf("  --help                Show this help message\n\n");
            printf("Examples:\n");
            printf("  %s 1234                 # Show memory info for PID 1234\n", argv[0]);
            printf("  %s 1234 --watch         # Monitor PID 1234\n", argv[0]);
            printf("  %s 1234 --watch 5       # Monitor with 5 sec interval\n", argv[0]);
            printf("  %s 1234 --compare 5678  # Compare two processes\n", argv[0]);
            printf("  %s 1234 --map           # Show memory map\n", argv[0]);
            exit(0);
        }
        else {
            fprintf(stderr, "Error: Unknown option '%s'\n", argv[i]);
            return -1;
        }
    }

    return 0;
}

int main(int argc, char* argv[]) {
    pid_t pid1, pid2;
    int watch_mode, interval, show_map;

    if (parse_arguments(argc, argv, &pid1, &pid2, &watch_mode, &interval, &show_map) != 0) {
        fprintf(stderr, "Usage: %s <PID> [--watch] [--interval SECONDS] [--compare PID2] [--map] [--help]\n", argv[0]);
        return 1;
    }

    signal(SIGINT, signal_handler);

    char proc_name[256];
    if (read_process_name(pid1, proc_name, sizeof(proc_name)) != 0) {
        fprintf(stderr, "Error: Process %d does not exist or access denied\n", pid1);
        return 1;
    }

    if (pid2 > 0) {
        ProcessInfo proc1, proc2;

        memset(&proc1, 0, sizeof(ProcessInfo));
        memset(&proc2, 0, sizeof(ProcessInfo));

        proc1.pid = pid1;
        proc2.pid = pid2;

        read_process_name(pid1, proc1.name, sizeof(proc1.name));
        read_process_name(pid2, proc2.name, sizeof(proc2.name));

        if (read_memory_metrics(pid1, &proc1.metrics) != 0) {
            fprintf(stderr, "Error: Cannot read memory metrics for PID %d\n", pid1);
            return 1;
        }

        if (read_memory_metrics(pid2, &proc2.metrics) != 0) {
            fprintf(stderr, "Error: Cannot read memory metrics for PID %d\n", pid2);
            return 1;
        }

        print_comparison(&proc1, &proc2);
        return 0;
    }

    if (watch_mode) {
        ProcessInfo proc;
        MemoryMetrics prev_metrics;
        int first_run = 1;

        memset(&proc, 0, sizeof(ProcessInfo));
        proc.pid = pid1;
        read_process_name(pid1, proc.name, sizeof(proc.name));

        printf("Monitoring %s (PID %d) - Update every %d seconds\n", proc.name, pid1, interval);
        printf("Press Ctrl+C to stop\n\n");

        print_watch_header();

        while (keep_running) {
            if (read_memory_metrics(pid1, &proc.metrics) != 0) {
                fprintf(stderr, "Error: Cannot read memory metrics for PID %d\n", pid1);
                break;
            }

            time_t now = time(NULL);
            struct tm* tm_info = localtime(&now);
            char time_str[32];
            strftime(time_str, sizeof(time_str), "%H:%M:%S", tm_info);

            printf("%-12s %-12s %-12s %-12s %-12ld %-12ld",
                time_str,
                format_size(proc.metrics.vms),
                format_size(proc.metrics.rss),
                proc.metrics.pss > 0 ? format_size(proc.metrics.pss) : "N/A",
                proc.metrics.minor_faults,
                proc.metrics.major_faults);

            if (!first_run) {
                char delta_buffer[32];
                printf(" | VSZ:%s RSS:%s",
                    format_delta(proc.metrics.vms - prev_metrics.vms, delta_buffer, sizeof(delta_buffer)),
                    format_delta(proc.metrics.rss - prev_metrics.rss, delta_buffer, sizeof(delta_buffer)));
            }

            printf("\n");
            fflush(stdout);

            prev_metrics = proc.metrics;
            first_run = 0;

            sleep(interval);
        }

        printf("\nMonitoring stopped.\n");
        return 0;
    }

    ProcessInfo proc;
    memset(&proc, 0, sizeof(ProcessInfo));
    proc.pid = pid1;

    if (read_process_name(pid1, proc.name, sizeof(proc.name)) != 0) {
        fprintf(stderr, "Error: Cannot read process name for PID %d\n", pid1);
        return 1;
    }

    if (read_memory_metrics(pid1, &proc.metrics) != 0) {
        fprintf(stderr, "Error: Cannot read memory metrics for PID %d\n", pid1);
        return 1;
    }

    print_basic_info(&proc);

    if (show_map) {
        if (read_memory_maps(pid1, &proc.maps, &proc.maps_count) == 0) {
            print_memory_map(&proc, 20);  
            free_memory_maps(proc.maps, proc.maps_count);
        }
        else {
            printf("\nCannot read memory map (access denied)\n");
        }
    }

    return 0;
}